Type annotations

davidatbu opened the pull request:
This PR annotates with types everything except `compat.py` and `tests/`.


Closes #42 .

davidatbu commented:
I am having trouble parsing the CI errors, would be happy to correct if you point me in the right direction @EmilStenstrom

davidatbu commented:
Ok, I think I've figued out why the CI is failing. The only thing that is failing is code coverage. And it's because of two changes that I made that testing doesn't cover:

```py
import typing as T

if T.TYPE_CHECKING:
   ...
```

This one is necessary to avoid cyclical imports _and_ annotate the argument type properly.

The other one is:

```py
class TokenTree:
    def __eq__(self, other: T.Any) -> bool:
        if isinstance(other, TokenTree):
            return self.token == other.token and self.children == other.children \
                and self.metadata == other.metadata
        return False
```

Before, the code assumed that `other` would also be of type `TokenTree`, but actually, `__eq__` needs to support comparing to other types of objects as well(mypy complains if one changes the annotation of `other` to `: 'TokenTree'`).

The second error can be addressed by adding a unit test. The first error, I'm not sure how to deal with, except for configuring the test coverage tool to not care about `if T.TYPE_CHECKING`, which I've no idea is possible or not.

EmilStenstrom commented:
The Typing error could be fixed by simply ignoring that line. I think it could be done by adding “  # pragma: no cover” after the typing if, or maybe adding if T.Typing here: https://github.com/EmilStenstrom/conllu/search?q=pragma&type=

davidatbu commented on conllu/models.py:
```diff
@@ -19,70 +35,71 @@ class Token(OrderedDict):
-    metadata = None
+class TokenList(T.List[Token]):
+
+    metadata: T.Optional[_MetadataType] = None
```
Is the class level declaration here meant to assign a default value of `None`, or just serve as documentation? If the latter is the case, then simply having `metadata: _MetadataType` would suffice here.

EmilStenstrom replied:
Yeah, this could probably be simplified. But since people could use this, changing it would break backwards compatibility, something I'd like to avoid. So let's keep it this way for now.

davidatbu commented on conllu/models.py:
```diff
@@ -19,70 +35,71 @@ class Token(OrderedDict):
-        if hasattr(iterable, 'metadata'):
-            if hasattr(self.metadata, '__add__') and hasattr(iterable.metadata, '__add__'):
+        if isinstance(iterable, TokenList):
+            if isinstance(self.metadata, Addable) and isinstance(iterable.metadata, Addable):
```
I introduced `Addable` "type checkable" way of checking if `self.metadata` has the `__add__` method. I actually wasn't clear on what exactly `__add__` is supposed to distinguish between(ie, what other types are valid types for `self.metadata`? list? tuple? str? All have `.__add__`). I think refactoring the code to be more explicit would make it clearer for future contributors.

EmilStenstrom replied:
It's supposed to mean (and is only tested for) "list of dicts". See the test_extend in tests/test_models.py. As long as that works, any simplifications is OK. If I understand things correctly this also means we can skip the typing_extensions dependency?

EmilStenstrom commented:
@davidatbu I've released version 4.0 of conllu, that drops support for anything < Python 3.6. If you rebase from latest master you should approach a shippable version of this PR.

EmilStenstrom commented:
Let's not add black right now, we can look at that after this fix has landed.

EmilStenstrom commented:
So next steps:

1. Rebase this branch on latest master
2. Fix any new typing errors reported by mypy
3. Add mypy checks to tox.ini and make sure they pass on CI
4. Refactor TokenList.extend to only support extending with TokenList or "Lists of Dicts" according to my comment above. 

Does this make sense?

EmilStenstrom commented:
@davidatbu Will you be taking on this work, or should I?

davidatbu commented:
Hi @EmilStenstrom , I apologize for the long silence. 

I rebased on master, refactored `TokenList.metadata` to be either `Metadata` or `List[Any]`(mypy doesn't support recursive types yet), and I fixed all the mypy errrors in the package, except one that I think is because of a bug. Here's the error:

```
/home/davidat/git/conllu/conllu/serializer.py:40:27: error: Item "List[Any]" of "Union[Metadata, List[Any]]" has no attribute "items"  [union-attr]
```
Indeed, if `TokenList.metadata` is allowed to be a list, the serializer method should accomodate that(or make it explicit that it doesn't accomodate that), right?

I don't have experience with tox, so I'd be glad if you could take over from here.

EmilStenstrom commented:
@davidatbu Thank you! I will take over from here.

EmilStenstrom commented:
Merged! Thank for contributing David.

I want to get a few more fixes in the next release, but I'll cut a new PyPI release soon.

EmilStenstrom commented:
I just released conllu 4.1 on pypi: https://pypi.org/project/conllu/ 🥳
